feat: slider value label (#275)
Atom-Learning/components: src/components/slider/Slider.mdx

---
title: Slider
component: Slider, Slider.Value, Slider.Steps
description: The slider component implements the Slider component from Radix, with default styling and the CSS prop.
category: Form primitives
---

The `Slider` is a simple component that renders a form slider.

> The `Slider` component should be rendered with a label for accesibility reasons. Rather than using the `Slider` component directly, it is best to use the `SliderField` component, which provides a label and easily integrates with the `Form` component. It also adds the `Slider.Value` by default (with options to remove) in order to create a unified experience without bloating the individual component.

Reviewer: ```suggestion
The `Slider` component should be rendered with a label for accesibility reasons. Rather than using the `Slider` component directly, it is best to use the `SliderField` component, which provides a field label, a value label, and easily integrates with the `Form` component.
```

Just a thought, to simplify a touch

<Slider defaultValue={[25, 75]} />
```

## Slider.Steps

A separate component exists to output an array of labels which appear at given value points along the slider. These are passed in to the `steps` property using an array objects that contain a label and a value.

This is most commonly used within the `SliderField`, but is here in case you need to compose your own complex component.

Note: it is likely better to create steps as a constant and pass in with `steps={steps}` or similar, but this preview code cannot see values outside of JSX.

```tsx preview
<Slider defaultValue={[50]}>
  <Slider.Steps
    min={0}
    max={100}
    steps={[
      { value: 0, label: 'min' },
      { value: 50, label: 'mid' },
      { value: 100, label: 'max' }
    ]}
  />
</Slider>
```

The component requires min and max values, these should be the same as the optional values that are passed to the Slider component.

`Slider.Steps` work well with the built in `step` property, which defaults to 1 and changes the size of each movement. For example, this would limit the slider to three values only.

> ```tsx preview

Reviewer: ```suggestion
```tsx
```

Let's not add too many rendered previews to the docs, a code example here is fine 👍

<Slider defaultValue={[50]} min={10} max={20} step={5}>
  <Slider.Steps
    min={10}
    max={20}
    steps={[
      { value: 10, label: 'min' },
      { value: 15, label: 'mid' },
      { value: 20, label: 'max' }
    ]}
  />
</Slider>
```

## Slider.Value

A separate component exists to output the value from a slider. This is most commonly used within the `SliderField`, but is here in case you need to compose your own complex component.

NOTE: this will only work if there is only one value selected. In order to not overly complicate the output, it is recommended to indicate multiple values using `steps` (see above). If more than one value is passed in, the component will not render.

```tsx preview
<Slider.Value value={[50]} />
```

> By default this simply outputs the value, however this can be customised in a number of ways. The property `outputLabel` accepts a callback function that takes a value and returns that in a string.

Reviewer: ```suggestion
By default this simply outputs the value, however, this can be customised in a number of ways. The property `outputLabel` accepts a function that passes the value selected in the slider and expects a string returned for the label.
```

```tsx preview
<Slider.Value
  value={[50]}
  outputLabel={(value) => `Current number of geese is ${value}`)}
/>

> ```

Reviewer: I'd remove this example and rely on the following code block to communicate the use of value and pluralisation.

You can also use this is in more complex ways, for example to set empty states and pluralisation, like below.

> ```tsx preview

Reviewer: I don't think a preview is useful here as the value can't change, so you won't be able to see the pluralisation

/>
```

## Value output

The slider can be set to display a label below the track displaying the current value selected. To do this, pass in the prop `showValue`.

NOTE: this will only work if there is only one value selected. In order to not overly complicate the output, it is recommended to indicate multiple values using `steps` (see above). Should you pass in this prop with multiple values, the value output will not be rendered.

```tsx preview

> <Slider defaultValue={[50]} showValue />

Reviewer 2: How come you're using an array here for the default value?

Author: values for the slider always have to be an array, even if there's only one. the slider extends the radix slider primitive and that's the format that uses. I know it reads a bit weird but it allows for multiple thumbs on the slider.

Reviewer 2: Ahh good to know - thx!

```

By default this simply outputs the value, however this can be customised in a number of ways. `outputLabel` can be passed, which is a string that must contain the `$VALUE`. This will be replaced with the value when it renders. A string is used so that the value can be placed wherever it needs to be in the output sentence, such as in the example below.

> The `outputLabel` also will turn the string `$S` into conditional plurals, i.e. 's' will be output if the value is greater than 1. This is also shown in the example review.

Reviewer 2: What happens if the plural is nonstandard, e.g., _child/children, pony/ponies, sheep/sheep_? Maybe we should have a prop to indicate what the plural form should be?

Author: fixed! great example of an edge case that's easily missed :)
